### PR TITLE
Migrating QML rendering off the main thread

### PR DESCRIPTION
**Pull request by jherico**
Push the rendering for each QML surface into it's own thread to avoid frequent OpenGL context switches on the main thread and allow QML rendering to never require a context switch on the target thread.

This is a significant change to the OffscreenQmlSurface internals.  All rendering logic is now encapsulated in a separate renderer class, so the class header itself doesn't need to declare as many members as before, simplifying it significantly.  

The design is based off the implementation of offscreen QML rendering as used in the Qt3D system, but with some improvments.

## Testing

Load the application verify correct rendering and interaction with WebOverlays (tablet, etc), WebEntities, and the desktop.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3938/HighFidelity-Beta-PR11537-5d3958a2bf4dad01488b5b1fe49d201fe6d031c9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3938/HighFidelity-Beta-PR11537-5d3958a2bf4dad01488b5b1fe49d201fe6d031c9.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3939/HighFidelity-Beta-PR11537-fded678195520f86a96053706d80ebc2101763e4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3939/HighFidelity-Beta-PR11537-fded678195520f86a96053706d80ebc2101763e4.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4206/HighFidelity-Beta-PR11537-c7e8cb379e6e18cc68035551af8a6a104e793d6a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4206/HighFidelity-Beta-PR11537-c7e8cb379e6e18cc68035551af8a6a104e793d6a.exe

**Comment by Nex-Pro** [comment]
T E S T I N G  Win10

**Comment by Nex-Pro** [comment]
### FAIL

I7 win10 gtx1070

Could not get this to stop crashing. Tried to create a web ent in HMD. invoked a crash on clicking the icon.
From there on I could not get interface to stay alive at local sandbox ( from this build )
Tried fresh install and in desktop to no avail.

My system was rebooted prior to this test.

Multiple splats submitted.

**Comment by jherico** [comment]
@Nex-Pro found an issue with rendering web entities.  Should be fixed now.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4308/HighFidelity-Beta-PR11537-e4801766d32c29a8468dad560657d396cc23efc6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4308/HighFidelity-Beta-PR11537-e4801766d32c29a8468dad560657d396cc23efc6.exe

**Comment by Nex-Pro** [comment]
T E S T I N G  Win10

**Comment by Nex-Pro** [comment]
I performed a full uninstall and removal of previous app data folders before installing this build.

Laser interaction with web entities is really poor. Either nothing happens or really slow scrolling. ( on the default HiFi page ) 
Using the stylus was really bad at first I had another user watch while I tried to use the menu. 
Typically it would select the menu item above the stylus. Took multiple attempts to set a checkbox.
Still have no lasers on HUD.  
Laser interaction with tablet buttons wasn't too bad. 
Do you want me to compare these behaviours with master?  @jherico

**Comment by Nex-Pro** [comment]
Oh and it isn't crashing !

**Comment by jherico** [comment]
@Nex-Pro this PR isn't intended to change the behavior of pointer interaction.  It's almost certainly the same as on master.  It's only an attempt to give more time to the main thread and to reduce the amount of time spent in global driver locks when rendering QML surfaces.

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -2161,12 +2163,28 @@ void Application::initializeGL() {
...
     }
 
+    // Build a shared canvas / context for the Chromium processes
     _glWidget->makeCurrent();
-    _chromiumShareContext = new OffscreenGLCanvas();
-    _chromiumShareContext->setObjectName("ChromiumShareContext");
-    _chromiumShareContext->create(_glWidget->qglContext());
-    _chromiumShareContext->makeCurrent();
-    qt_gl_set_global_share_context(_chromiumShareContext->getContext());
+    if (!nsightActive()) {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: what happens when nsight is active? do web pages just silently fail? maybe we should include a qWarning() or something to indicate why.

Comment (jherico) [reply interface/src/Application.cpp]: added a warning and a comment to the code.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4385/HighFidelity-Beta-PR11537-3ef7cb7e0f1ea1d6c1fc251bcf200089dea7a891.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4385/HighFidelity-Beta-PR11537-3ef7cb7e0f1ea1d6c1fc251bcf200089dea7a891.exe

**Comment by samcake** [comment]
We need a test plan for this in test rail
we MUST test on i5 and verufy that  the performances are correct

**Comment by Nex-Pro** [comment]
@jherico Apologies, I only just saw your comment of 21 days ago.

**Comment by Nex-Pro** [comment]
Retesting

**Comment by Nex-Pro** [comment]
### Fail

QML keyboard does not work when inputting characters on a web page. ( works on master ) 

To repro: 

Open the create app in the tablet.
Create a web entity.
Scroll the high fidelity page until you see the download button.
Click the download button.
Click the steam button.
Click and gain cursor focus in the search box near the top of the steam page.
Observe that the QML keyboard is raised.
Type some characters in the search box.
Observe that nothing happens.
( In master the chars you type appear in the search box ).

**Comment by Nex-Pro** [comment]
Unsure if the above is related to this PR. or due to being outdated.

**Comment by conklin94122** [comment]
Does this need a milestone, @divyaelango88 ?

**Comment by divyaelango88** [comment]
no John it does not. Its an improvement, does not have a deadline associated with it

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5722/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5722/HighFidelity-Beta-PR11537-8aff405048ab3e794abc77ffe37725c43bbd7e0c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5722/HighFidelity-Beta-PR11537-8aff405048ab3e794abc77ffe37725c43bbd7e0c.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5748/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5748/HighFidelity-Beta-PR11537-a809f9846ca54735bb53ccdc8fce2c853ef133b7.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5748/HighFidelity-Beta-PR11537-a809f9846ca54735bb53ccdc8fce2c853ef133b7.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5997/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6012/app_11537_PR.apk

**Comment by conklin94122** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6023/app_11537_PR.apk

**Comment by conklin94122** [comment]
@jherico This has failed to build a couple times. Can you take a peek, please? Marking as QA Failed for now.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6044/app_11537_PR.apk

**Comment by conklin94122** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6066/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6078/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6078/HighFidelity-Beta-PR11537-c8cb854054433d49d868b75ba428441e1970007a.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6082/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6082/HighFidelity-Beta-PR11537-6c54f75eb2f77e79b104c1b1605938883182ff12.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6082/HighFidelity-Beta-PR11537-6c54f75eb2f77e79b104c1b1605938883182ff12.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6087/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6087/HighFidelity-Beta-PR11537-c1f525bad565122f8c6ce656c9299fef6158b9b7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6087/HighFidelity-Beta-PR11537-c1f525bad565122f8c6ce656c9299fef6158b9b7.exe

**Comment by cain-kilgore** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6096/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6096/HighFidelity-Beta-PR11537-c1f525bad565122f8c6ce656c9299fef6158b9b7.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6096/HighFidelity-Beta-PR11537-c1f525bad565122f8c6ce656c9299fef6158b9b7.dmg

**Comment by conklin94122** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6151/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6151/HighFidelity-Beta-PR11537-c1f525bad565122f8c6ce656c9299fef6158b9b7.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6151/HighFidelity-Beta-PR11537-c1f525bad565122f8c6ce656c9299fef6158b9b7.dmg

**Comment by davidkelly** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6171/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6171/HighFidelity-Beta-PR11537-c1f525bad565122f8c6ce656c9299fef6158b9b7.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6171/HighFidelity-Beta-PR11537-c1f525bad565122f8c6ce656c9299fef6158b9b7.dmg

**Comment by conklin94122** [comment]
@jherico We keep getting build failures on this PR. Please take a look.

**Comment by conklin94122** [comment]
@jherico Please add QA Ready when it's behaving.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6330/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6330/HighFidelity-Beta-PR11537-d3d45b5f2baa60cfb8afdf09234cdec90e6f5e18.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6330/HighFidelity-Beta-PR11537-d3d45b5f2baa60cfb8afdf09234cdec90e6f5e18.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6333/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6333/HighFidelity-Beta-PR11537-cfd4ce4e0a9c22650c1953a80529350cc35431a6.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6333/HighFidelity-Beta-PR11537-cfd4ce4e0a9c22650c1953a80529350cc35431a6.dmg

**Comment by jherico** [comment]
@conklin94122 build is fixed.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6355/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6355/HighFidelity-Beta-PR11537-216952d22ebeee86dc637756e62204877c0a4d7c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6355/HighFidelity-Beta-PR11537-216952d22ebeee86dc637756e62204877c0a4d7c.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6356/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6356/HighFidelity-Beta-PR11537-216952d22ebeee86dc637756e62204877c0a4d7c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6356/HighFidelity-Beta-PR11537-216952d22ebeee86dc637756e62204877c0a4d7c.dmg

**Comment by sbennett77** [comment]
Currently Testing.

**Comment by sbennett77** [comment]
Failed. Web browser would not load in desktop or VR (see screenshots). All other tablet actions seemed to function normal.

@conklin94122 


blank web
vr blank web

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6957/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6957/HighFidelity-Beta-PR11537-7d5502eb929bd4347824992df3a17fe7748109b7.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7012/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7012/HighFidelity-Beta-PR11537-fcb59a4ee6f9242aef1825fbdbf74a85e13f5307.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7012/HighFidelity-Beta-PR11537-fcb59a4ee6f9242aef1825fbdbf74a85e13f5307.exe

**Comment by sbennett77** [comment]
Currently Testing.

**Comment by sbennett77** [comment]
Passed.

@conklin94122

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7088/app_11537_PR.apk

**Comment by jherico** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7097/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7101/app_11537_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7101/HighFidelity-Beta-PR11537-7102cf44b87cd54fa328f030d768d225b212da95.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7101/HighFidelity-Beta-PR11537-7102cf44b87cd54fa328f030d768d225b212da95.exe

**Pull request merged**